File: mobile/ios/Classes/CouchPersistentReplicationProxy.m
//
//  CouchPersistentReplicationProxy.m
//  titouchdb
//
//  Created by Paul Mietz Egli on 4/5/12.
//  Copyright (c) 2012 __MyCompanyName__. All rights reserved.
//

#import "CouchPersistentReplicationProxy.h"
#import "TiProxy+Errors.h"

@implementation CouchPersistentReplicationProxy

@synthesize replication;

- (id)initWithCouchPersistentReplication:(CouchPersistentReplication *)rep {
    if (self = [super init]) {
        self.replication = rep;
    }
    return self;
}

+ (CouchPersistentReplicationProxy *)proxyWith:(CouchPersistentReplication *)rep {
    return rep ? [[[CouchPersistentReplicationProxy alloc] initWithCouchPersistentReplication:rep] autorelease] : nil;
}

/* TODO
- (id)source {
    return self.replication.source;
}

- (id)target {
    return self.replication.target;
}
*/

- (id)remoteURL {
    return [self.replication.remoteURL absoluteString];
}

- (id)createTarget {
    return [NSNumber numberWithBool:self.replication.create_target];
}

- (void)setCreateTarget:(id)val {
    self.replication.create_target = [val boolValue];
}

- (id)continuous {
    return [NSNumber numberWithBool:self.replication.continuous];
}

- (void)setContinuous:(id)val {
    self.replication.continuous = [val boolValue];
}

- (id)filter {
    return self.replication.filter;
}

- (void)setFilter:(id)val {
    self.replication.filter = val;
}

- (id)filterParams {
    return self.replication.query_params;
}

- (id)setFilterParams:(id)val {
    self.replication.query_params = val;
}

- (void)actAsUserWithRoles:(id)args {
    NSString * username;
    NSArray * roles;
    
    ENSURE_ARG_OR_NIL_AT_INDEX(username, args, 0, NSString)
    ENSURE_ARG_OR_NIL_AT_INDEX(roles, args, 1, NSArray)
    
    [self.replication actAsUser:username withRoles:roles];
}

- (void)actAsAdmin:(id)args {
    [self.replication actAsAdmin];
}

- (void)restart:(id)args {
    [self.replication restart];
}

- (id)status {
    return [NSNumber numberWithInt:self.replication.state];
}

- (id)completed {
    return [NSNumber numberWithBool:self.replication.completed];
}

- (id)total {
    return [NSNumber numberWithUnsignedInt:self.replication.total];
}

- (id)error {
    return [self errorDict:self.replication.error];
}

- (id)mode {
    return [NSNumber numberWithInt:self.replication.mode];
}

@end

File: mobile/android/src/com/obscure/titouchdb/old/CouchPersistentReplicationProxy.java
package com.obscure.titouchdb.old;

import java.util.HashMap;
import java.util.Map;
import java.util.Observable;
import java.util.Observer;

import org.appcelerator.kroll.KrollDict;
import org.appcelerator.kroll.KrollProxy;
import org.appcelerator.kroll.annotations.Kroll;

import com.couchbase.touchdb.TDDatabase;
import com.couchbase.touchdb.replicator.TDReplicator;
import com.obscure.titouchdb.TitouchdbModule;

/**
 * TODO this isn't really a persistent replication -- it just fakes it using a
 * regular replication object.
 * 
 * @author Paul Mietz Egli
 * 
 */
@Kroll.proxy(parentModule = TitouchdbModule.class)
public class CouchPersistentReplicationProxy extends KrollProxy implements Observer {

    private static final String LCAT                        = "CouchPersistentReplicationProxy";

    private static final String REPLICATOR_PROGRESS_CHANGED = "progress";

    private static final String REPLICATOR_PROGRESS_STOPPED = "stopped";

    private TDDatabase          db;

    private TDReplicator        repl;

    public CouchPersistentReplicationProxy(TDDatabase db, TDReplicator repl) {
        assert db != null;
        assert repl != null;
        this.db = db;
        this.repl = repl;
        repl.addObserver(this);
    }

    @Kroll.method
    public void actAsAdmin() {
        // TODO
    }

    @Kroll.method
    public void actAsUserWithRoles(String username, String[] roles) {
        // TODO
    }

    @Kroll.getProperty(name = "error")
    public KrollDict getError() {
        // TODO
        return null;
    }

    @Kroll.getProperty(name = "filter")
    public String getFilter() {
        // TODO
        return null;
    }

    @Kroll.getProperty(name = "filterParams")
    public KrollDict getFilterParams() {
        // TODO
        return null;
    }

    @Kroll.getProperty(name = "mode")
    public int getMode() {
        // TODO get actual mode
        if (repl.isRunning()) {
            return TitouchdbModule.REPLICATION_MODE_ACTIVE;
        }
        else {
            return TitouchdbModule.REPLICATION_MODE_STOPPED;
        }
    }

    @Kroll.getProperty(name = "remoteURL")
    public String getRemoteURL() {
        return repl.getRemote().toString();
    }

    @Kroll.getProperty(name = "source")
    public String getSource() {
        return repl.isPush() ? db.getName() : repl.getRemote().toString();
    }

    @Kroll.getProperty(name = "status")
    public int getStatus() {
        // TODO real status
        return 0;
        /*
        if (getCompleted() == getTotal()) {
            return TitouchdbModule.REPLICATION_STATE_COMPLETED;
        }
        else {
            return TitouchdbModule.REPLICATION_STATE_IDLE;
        }
        */
    }

    @Kroll.getProperty(name = "target")
    public String getTarget() {
        return repl.isPush() ? repl.getRemote().toString() : db.getName();
    }

    @Kroll.getProperty(name = "total")
    public int getTotal() {
        return repl.getChangesTotal();
    }

    @Kroll.getProperty(name = "completed")
    public int getCompleted() {
        return repl.getChangesProcessed();
    }

    @Kroll.getProperty(name = "continuous")
    public boolean isContinuous() {
        // TODO
        return false;
    }

    @Kroll.method
    public void restart() {
        // TODO
        repl.stop();
        repl.start();
    }

    @Kroll.setProperty(name = "filter")
    public void setFilter(String filter) {
        // TODO
    }

    @Kroll.setProperty(name = "filterParams")
    public void setFilterParams(KrollDict params) {
        // TODO
    }

    private Map<String, Object> toStatusDictionary() {
        Map<String, Object> result = new HashMap<String, Object>();
        result.put("completed", repl.getChangesProcessed());
        result.put("total", repl.getChangesTotal());
        result.put("status", getStatus());
        result.put("error", getError());
        result.put("running", repl.isRunning());
        return result;
    }

    public void update(Observable observable, Object arg) {
        Map<String, Object> dict = toStatusDictionary();
        if (!repl.isRunning()) {
            this.fireEvent(REPLICATOR_PROGRESS_STOPPED, dict);
        }
        else {
            this.fireEvent(REPLICATOR_PROGRESS_CHANGED, dict);
        }
    }
}
